### ISSUE TITLE
Unable to perform backup from standby server to S3 storage (ceph rados gateway)

### ISSUE THREAD

**Issue by mt-kaslo** [pgbackrest/pgbackrest#1259]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.3

2. PostgreSQL version: 9.6

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
Ubuntu 16.04.6 LTS (all systems)

4. Did you install pgBackRest from source or from a package? Yes, apt install

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
    - errors in the postgresql log file before or during the time you experienced the issue
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

7. Describe the issue:
Configuring pgBackRest to backup from the standby server fails.
---
I have configured pgBackRest to backup from the primary server to a ceph rados gateway that emulates Amazon's S3 interface. I have successfully created both a full and a differential backup into a bucket.
```
sudo -u postgres pgbackrest --stanza=9-6-main info
stanza: 9-6-main
    status: ok
    cipher: none

    db (current)
        wal archive min/max (9.6-1): 0000002A0000387D00000010/0000002A0000387D00000022

        full backup: 20201211-232749F
            timestamp start/stop: 2020-12-11 23:27:49 / 2020-12-11 23:31:15
            wal start/stop: 0000002A0000387D00000012 / 0000002A0000387D00000012
            database size: 301MB, backup size: 301MB
            repository size: 30.8MB, repository backup size: 30.8MB

        diff backup: 20201211-232749F_20201214-223958D
            timestamp start/stop: 2020-12-14 22:39:58 / 2020-12-14 22:40:12
            wal start/stop: 0000002A0000387D00000019 / 0000002A0000387D00000019
            database size: 301.4MB, backup size: 103.6MB
            repository size: 30.9MB, repository backup size: 15.7MB
            backup reference list: 20201211-232749F
```

I have worked through the the User Guide and setup Passwordless SSH:
[Setup Passwordless SSH](https://pgbackrest.org/user-guide.html#repo-host/setup-ssh)

This has been tested and works.

PostgreSQL was previously configured as a Hot/Standby system, so the only additional lines I needed to add to the postgresql.conf file were:
```
archive_mode = on
archive_command = 'pgbackrest --stanza=9-6-main archive-push %p'
```
The other parameters in the file were set appropriately:
```
listen_addresses = '*'
max_wal_senders = 3
wal_level = hot_standby
```
Both the primary PostgreSQL and standby server are configured identically. The WAL archives are being successfully pushed to the S3 storage (ceph rados gateway).

The pgbackrest.conf file on the primary system that allows successful backups is:
```
[9-6-main]
pg1-host=primary-2.server
pg1-path=/var/lib/postgresql/9.6/main
pg1-port=<some-port>

[global]
process-max=4
log-level-console=info
log-level-file=debug
# Full backups are performed monthly
# Setting the number of days to 65 ensures at least 2 monthly backups are retained
repo1-retention-full-type=time
repo1-retention-full=65
# Differential backups are performed weekly
# Setting the count to 4 ensures at least 3 weekly backups
repo1-retention-diff=4
repo1-path=/postgresql
repo1-type=s3
repo1-s3-bucket=pgbackrest
repo1-s3-endpoint=rados-gateway.server
repo1-s3-region=default
repo1-s3-uri-style=path
repo1-s3-port=<another-port>
repo1-s3-key=<redacted>
repo1-s3-key-secret=<redacted>
repo1-s3-verify-tls=n

[global:archive-push]
compress-level=3
```
Changing the file to:
```
[9-6-main]
pg1-host=primary-2.server
pg1-path=/var/lib/postgresql/9.6/main
pg1-port=<some-port>
pg2-host=standby-1.server
pg2-host-user=pgbackrest
pg2-path=/var/lib/postgresql/9.6/main
pg2-port=<some-port>

[global]
backup-standby=y
process-max=4
log-level-console=info
log-level-file=debug
# Full backups are performed monthly
# Setting the number of days to 65 ensures at least 2 monthly backups are retained
repo1-retention-full-type=time
repo1-retention-full=65
# Differential backups are performed weekly
# Setting the count to 4 ensures at least 3 weekly backups
repo1-retention-diff=4
repo1-path=/postgresql
repo1-type=s3
repo1-s3-bucket=pgbackrest
repo1-s3-endpoint=rados-gateway.server
repo1-s3-region=default
repo1-s3-uri-style=path
repo1-s3-port=<another-port>
repo1-s3-key=<redacted>
repo1-s3-key-secret=<redacted>
repo1-s3-verify-tls=n

[global:archive-push]
compress-level=3
```
... and running the command:
'sudo -u postgres pgbackrest --stanza=9-6-main --type=incr backup > /tmp/9-6-main-incr2.log 2>&1

results in the following errors:
```
2020-12-15 23:08:57.411 P00   INFO: backup command begin 2.30: --backup-standby --log-level-console=info --log-level-file=debug --pg2-host=standby-1.server --pg2-host-user=pgbackrest --pg1-local --pg1-path=/var/lib/postgresql/9.6/main --pg2-path=/var/lib/postgresql/9.6/main --pg1-port=<some-port> --pg2-port=<some-port> --process-max=4 --repo1-path=/postgresql --repo1-retention-diff=4 --repo1-retention-full=65 --repo1-retention-full-type=time --repo1-s3-bucket=pgbackrest --repo1-s3-endpoint=rados-gateway.server --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-port=<another-port> --repo1-s3-region=default --repo1-s3-uri-style=path --no-repo1-s3-verify-tls --repo1-type=s3 --stanza=9-6-main --type=incr
WARN: unable to check pg-2: [LockAcquireError] raised from remote-0 protocol on 'standby-1.server': raised from remote-0 protocol on 'primary-2.server': unable to acquire lock on file '/tmp/pgbackrest/9-6-main-backup.lock': Resource temporarily unavailable
      HINT: is another pgBackRest process running?
      protocol/client:protocolClientProcessError:177:(trace log level required for parameters)
      protocol/client:protocolClientReadOutput:(trace log level required for parameters)
      protocol/client:protocolClientExecute:(trace log level required for parameters)
      protocol/client:protocolClientNoOp:(trace log level required for parameters)
      protocol/client:protocolClientNew:(trace log level required for parameters)
      protocol/helper:protocolRemoteGet:(debug log level required for parameters)
      storage/remote/protocol:storageRemoteProtocol:(debug log level required for parameters)
      protocol/server:protocolServerProcess:(debug log level required for parameters)
      command/remote/remote:cmdRemote:(debug log level required for parameters)
      main:main:(debug log level required for parameters)
ERROR: [056]: unable to find standby cluster - cannot proceed
2020-12-15 23:08:59.792 P00   INFO: backup command end: aborted with exception [056]
```

Standby pgbackrest.conf is:
```
[9-6-main]
pg1-host=primary-2.server
pg1-path=/var/lib/postgresql/9.6/main
pg1-port=<some-port>
pg2-host=standby-1.server
pg2-path=/var/lib/postgresql/9.6/main
pg2-port=<some-port>

[global]
backup-standby=y
process-max=4
log-level-console=info
log-level-file=debug
# Full backups are performed monthly
# Setting the number of days to 65 ensures at least 2 monthly backups are retained
repo1-retention-full-type=time
repo1-retention-full=65
# Differential backups are performed weekly
# Setting the count to 4 ensures at least 3 weekly backups
repo1-retention-diff=4
# Repository settings
repo1-path=/postgresql
repo1-type=s3
repo1-s3-bucket=pgbackrest
repo1-s3-endpoint=rados-gateway.server
repo1-s3-region=default
repo1-s3-uri-style=path
repo1-s3-port=<another-port>
repo1-s3-key=<redacted>
repo1-s3-key-secret=<redacted>
repo1-s3-verify-tls=n

[global:archive-push]
compress-level=3
```

I have looked at issues #1219 and #1197 and the issues I am encountering is extremely similar to these. I used the advice to create a /etc/pgbackrest/conf.d/local.conf on each of the PostgreSQL servers, but this did not seem to have any effect.

primary-2.server /etc/pgbackrest/conf.d/local.conf:
```
[9-6-main]
pg1-local=y
```
standby-1.server /etc/pgbackrest/conf.d/local.conf:
```
[9-6-main]
pg2-local=y
```

I need advice on how to configure the primary and standby server pgBackRest.conf files to allow backup of the PostgreSQL database from the standby server to the S3 storage.

Thanks in advance.

**Comment by pgstef** [comment]
Hi,

The issue #1197 is a good example. It states that currently, you can have symmetrical configurations.
Then, don't specify the "host" for your locally running PG cluster.

Given the info you provided, I think adjusting your settings like the following should help:

```ini
# primary
[9-6-main]
pg1-path=/var/lib/postgresql/9.6/main
pg1-port=<some-port>
pg2-host=standby-1.server
pg2-host-user=pgbackrest
pg2-path=/var/lib/postgresql/9.6/main

# standby
[9-6-main]
pg1-host=primary-2.server
pg1-host-user=pgbackrest (?)
pg1-path=/var/lib/postgresql/9.6/main
pg2-path=/var/lib/postgresql/9.6/main
pg2-port=<some-port>
```

On both servers you're then using the `pgbackrest` local user to run the pgbackrest commands and have password-less ssh connections on both directions. Right ?

With the above settings, you should be able to run the backup command (using `--backup-standby` or not) from both servers. I'm then just wondering why you're trying to actually run the backup command on the primary server if you want to backup from the standby. You could simply run it from the standby server itself to avoid the load on the primary.

**Comment by mt-kaslo** [comment]
Thanks for the reply and advice @pgstef.

Your advice has me moving down the right path. The [Setup Passwordless SSH](https://pgbackrest.org/user-guide.html#repo-host/setup-ssh) shows the generation of ssh keys for the `pgbackrest` user on the repository and ssh keys for the `postgres` user on the primary. I also mixed in the instructions from Crunchy Data, [pgBackRest - Performing Backups on a Standby Cluster](https://info.crunchydata.com/blog/pgbackrest-performing-backups-on-a-standby-cluster). The section, _Configure Passwordless SSH_, is very similar to the pgbackrest instructions in that it also shows the generation of ssh keys for the `pgbackrest` user on the replica (standby) host and ssh keys for the `postgres` user on the primary system.

I believe have different users (pgbackrest on the standby, postgres on the primary) is causing issues with connecting to the database on the primary system.

I have made the changes you suggested to my `pgbackrest.conf` on both the primary and standby systems.

Running the following command on the standby system:
`sudo -u pgbackrest pgbackrest --stanza=9-6-main --type=incr backup > /tmp/9-6-main-incr3.log 2>&1`
results in the following error message:
> 2020-12-16 17:21:10.158 P00   INFO: backup command begin 2.30: --backup-standby --log-level-console=info --log-level-file=debug --pg1-host=postgresql.kin --pg1-host-user=pgbackrest --pg2-local --pg1-path=/var/lib/postgresql/9.6/main --pg2-path=/var/lib/postgresql/9.6/main --pg1-port=5432 --pg2-port=5432 --process-max=4 --repo1-path=/postgresql --repo1-retention-diff=4 --repo1-retention-full=65 --repo1-retention-full-type=time --repo1-s3-bucket=pgbackrest --repo1-s3-endpoint=ceph-rgw.kin --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-port=7443 --repo1-s3-region=default --repo1-s3-uri-style=path --no-repo1-s3-verify-tls --repo1-type=s3 --stanza=9-6-main --type=incr
WARN: unable to check pg-1: [UnknownError] remote-0 process on 'postgresql.kin' terminated unexpectedly [255]: Permission denied (publickey).
WARN: unable to check pg-2: [DbConnectError] unable to connect to 'dbname='postgres' port=5432': **FATAL:  role "pgbackrest" does not exist**
ERROR: [056]: unable to find primary cluster - cannot proceed
2020-12-16 17:21:10.387 P00   INFO: backup command end: aborted with exception [056]

Running the following command on the standby system:
`sudo -u postgres pgbackrest --stanza=9-6-main --type=incr backup > /tmp/9-6-main-incr4.log 2>&1`
results in the following error message:

> WARN: unable to open log file '/var/log/pgbackrest/9-6-main-backup.log': Permission denied
      NOTE: process will continue without log file.
2020-12-16 17:29:20.375 P00   INFO: backup command begin 2.30: --backup-standby --log-level-console=info --log-level-file=off --pg1-host=postgresql.kin --pg1-host-user=pgbackrest --pg2-local --pg1-path=/var/lib/postgresql/9.6/main --pg2-path=/var/lib/postgresql/9.6/main --pg1-port=5432 --pg2-port=5432 --process-max=4 --repo1-path=/postgresql --repo1-retention-diff=4 --repo1-retention-full=65 --repo1-retention-full-type=time --repo1-s3-bucket=pgbackrest --repo1-s3-endpoint=ceph-rgw.kin --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-port=7443 --repo1-s3-region=default --repo1-s3-uri-style=path --no-repo1-s3-verify-tls --repo1-type=s3 --stanza=9-6-main --type=incr
WARN: unable to check pg-1: [UnknownError] remote-0 process on 'postgresql.kin' terminated unexpectedly [255]: <b>Permission denied (publickey).</b>
ERROR: [056]: unable to find primary cluster - cannot proceed
2020-12-16 17:29:21.209 P00   INFO: backup command end: aborted with exception [056]

I added emphasis to the error messages to highlight the differences. 

In the first instance, user `pgbackrest` is able to connect via ssh, **but** does not have a role configured in the PostgreSQL 
database.

In the second instance, user `postgres` does **note** have proper ssh credentials configured and is unable to connect to the primary server.

I need to sort out and quite likely create a role in the database for backups and assign that to a database user.

It's amazing what a good nights sleep, coffee and good advice (thanks @pgstef) can do!

**Comment by pgstef** [comment]
You only need the SSH connection because you try to take the backup from the primary server, then connect to the standby server, read the data directory there, and push it to the S3 repo from the primary server. That's usually not the point of the `backup-standby` option.

You should probably run the backup command directly on the standby server. That wouldn't require any SSH connection then.
The ssh connection is usually configured when a dedicated remote backup server is created (without any PG cluster running on it). That's why we advise to create a `pgbackrest` system user there. 

In your case, creating a `pgbackrest` db role doesn't seems like the thing to do. My best advice would be to trigger the backup command directly on the standby server. And if you really want to trigger the backup from the primary server, then configure the ssh connections for the `postgres` system user.

[This](https://pgstef.github.io/2020/09/04/combining_pgbackrest_and_streaming_replication.html) (old) blog post is pretty much showing what you try to achieve (but you'll store the repo on S3 rather than NFS).

**Comment by mt-kaslo** [comment]
@pgstef Sorry I didn't make it clear that the `backup-standby=y` **is** configured in the `pgbackrest.conf` file on both the primary and standby servers.

I believe I am triggering the backup by running the pgbackrest command on the standby server **only**.

The primary and standby PostgreSQL servers are running a Hot/Standby configuration (`wal_level = hot_standby` is configured in the postgresql.conf file on both servers).

I haven't seen any configuration to instruct pgBackRest on the standby server **not** to contact the primary server via ssh.

I have some more reading to do (thanks for the link), thinking and experimenting.

**Comment by mt-kaslo** [comment]
I have a working solution!

I generated ssh keys for the `postgres` user on the standby-1.server and installed the public key into the `postgres` user on the primary-2.server. I have also installed the public key for the `postgres` user on the primary-2.server onto the standby-1.server.

Backups can now run from the command-line on the standby server and it appears the majority of the PostgreSQL files are copied from the standby-1.server to our S3 bucket (ceph rados gateway). The majority of the backup load is on the standby-1.server and **not** on the primary-2.server.

Remaining testing is to failover from the primary-2.server to the standby-1.server and ensure pgbackrest runs on the primary-2.server.
